### ISSUE TITLE
Data::Dumper of regexps not idempotent

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12510]
Migrated from [rt.perl.org#115430](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=115430) (status was 'resolved')

Searchable as RT115430$

**Comment by p5pRT** [comment]
### From @epa

#### Created by @epa



Data&#8203;::Dumper&#8203;::Dumper gives a string of code you can eval\(\) to get
back the original value\.  But when dumping a regular expression\,
some additional ?^u stuff is added to the regexp\.  That would be
fine\, except that it gets added every time even if already present\.
Hence dumping and evaling a regexp adds another ?^u every time\.

The dump\-eval cycle may not give back the exact same string of
code you started with \- it's fine for it to translate things to a
canonical representation \- but it should be idempotent\.  Evaling
a dumped value and then dumping it again should give the same thing\.

&nbsp;    use 5\.014;
&nbsp;    use warnings;
&nbsp;    use strict;
&nbsp;    use Data&#8203;::Dumper qw\(Dumper\);
&nbsp;    our $VAR1;

&nbsp;    my $re = qr/abc/;
&nbsp;    my $d = Dumper $re;
&nbsp;    my $re2 = eval $d;
&nbsp;    my $d2 = Dumper $re2;

&nbsp;    say $d;
&nbsp;    say $d2;

I expected both lines to be the same but they differ&#8203;:

$VAR1 = qr/\(?^u&#8203;:abc\)/;

$VAR1 = qr/\(?^u&#8203;:\(?^u&#8203;:abc\)\)/;

FWIW\, the non\-core Data&#8203;::Dump module does the same thing\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

This perlbug was built using Perl 5.14.2 in the Fedora build system.
It is being executed now by Perl 5.14.2 - Fri Sep 14 12:06:01 UTC 2012.

Site configuration information for perl 5.14.2:

Configured by Red Hat, Inc. at Fri Sep 14 12:06:01 UTC 2012.

Summary of my perl5 (revision 5 version 14 subversion 2) configuration:
   
  Platform:
    osname=linux, osvers=2.6.32-279.2.1.el6.x86_64,
archname=x86_64-linux-thread-multi
    uname='linux buildvm-04.phx2.fedoraproject.org
2.6.32-279.2.1.el6.x86_64 #1 smp thu jul 5 21:08:58 edt 2012 x86_64
x86_64 x86_64 gnulinux '
    config_args='-des -Doptimize=-O2 -g -pipe -Wall
-Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector
--param=ssp-buffer-size=4  -m64 -mtune=generic
-Dccdlflags=-Wl,--enable-new-dtags -DDEBUGGING=-g -Dversion=5.14.2
-Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red
Hat, Inc. -Dprefix=/usr -Dvendorprefix=/usr -Dsiteprefix=/usr/local
-Dsitelib=/usr/local/share/perl5 -Dsitearch=/usr/local/lib64/perl5
-Dprivlib=/usr/share/perl5 -Dvendorlib=/usr/share/perl5/vendor_perl
-Darchlib=/usr/lib64/perl5 -Dvendorarch=/usr/lib64/perl5/vendor_perl
-Darchname=x86_64-linux-thread-multi -Dlibpth=/usr/local/lib64 /lib64
/usr/lib64 -Duseshrplib -Dusethreads -Duseithreads
-Dusedtrace=/usr/bin/dtrace -Duselargefiles -Dd_semctl_semun -Di_db
-Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio
-Dinstallusrbinperl=n -Ubincompat5005 -Uversiononly
-Dpager=/usr/bin/less -isr -Dd_gethostent_r_proto -Ud_endhostent_r_proto
-Ud_sethostent_r_proto -Ud_endprotoent_r_proto -Ud_setprotoent_r_proto
-Ud_endservent_r_proto -Ud_setservent_r_proto -Dscriptdir=/usr/bin'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing
-pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE
-D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions
-fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe
-fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.6.3 20120306 (Red Hat 4.6.3-2)',
gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector'
    libpth=/usr/local/lib64 /lib64 /usr/lib64
    libs=-lresolv -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread
-lc -lgdbm_compat
    perllibs=-lresolv -lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.14.90'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef,
ccdlflags='-Wl,--enable-new-dtags -Wl,-rpath,/usr/lib64/perl5/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -pipe -Wall
-Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector
--param=ssp-buffer-size=4 -m64 -mtune=generic'

Locally applied patches:
    


@INC for perl 5.14.2:
    /home/eda/lib/perl5//x86_64-linux-thread-multi
    /home/eda/lib/perl5/
    /usr/local/lib64/perl5
    /usr/local/share/perl5
    /usr/lib64/perl5/vendor_perl
    /usr/share/perl5/vendor_perl
    /usr/lib64/perl5
    /usr/share/perl5
    .


Environment for perl 5.14.2:
    HOME=/home/eda
    LANG=en_GB.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=C
    LC_CTYPE=en_GB.UTF-8
    LC_MESSAGES=en_GB.UTF-8
    LC_MONETARY=en_GB.UTF-8
    LC_NUMERIC=en_GB.UTF-8
    LC_TIME=en_GB.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
 
PATH=/home/eda/bin:/home/eda/bin:/usr/local/bin:/bin:/usr/bin:/sbin:/usr
/sbin:/sbin:/usr/sbin
    PERL5LIB=/home/eda/lib/perl5/
    PERL_BADLANG (unset)
    SHELL=/bin/bash


______________________________________________________________________
This email has been scanned by the Symantec Email Security.cloud service.
For more information please visit http://www.symanteccloud.com
______________________________________________________________________

```
</details>

**Comment by p5pRT** [comment]
### From @b2gills

On Tue\, Oct 23\, 2012 at 7&#8203;:28 AM\, Ed Avis \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  "Ed Avis"
> \# Please include the string&#8203;:  \[perl \#115430\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=115430 >
>
>
> This is a bug report for perl from eda@&#8203;waniasset\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.14\.2\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
> Data&#8203;::Dumper&#8203;::Dumper gives a string of code you can eval\(\) to get
> back the original value\.  But when dumping a regular expression\,
> some additional ?^u stuff is added to the regexp\.  That would be
> fine\, except that it gets added every time even if already present\.
> Hence dumping and evaling a regexp adds another ?^u every time\.
>
> The dump\-eval cycle may not give back the exact same string of
> code you started with \- it's fine for it to translate things to a
> canonical representation \- but it should be idempotent\.  Evaling
> a dumped value and then dumping it again should give the same thing\.
>
>     use 5\.014;
>     use warnings;
>     use strict;
>     use Data&#8203;::Dumper qw\(Dumper\);
>     our $VAR1;
>
>     my $re = qr/abc/;
>     my $d = Dumper $re;
>     my $re2 = eval $d;
>     my $d2 = Dumper $re2;
>
>     say $d;
>     say $d2;
>
> I expected both lines to be the same but they differ&#8203;:
>
> $VAR1 = qr/\(?^u&#8203;:abc\)/;
>
> $VAR1 = qr/\(?^u&#8203;:\(?^u&#8203;:abc\)\)/;

Add&#8203;:

&nbsp;  say $re;
&nbsp;  say $re2;

then it prints

&nbsp;  \(?^u&#8203;:abc\)
&nbsp;  \(?^u&#8203;:\(?^u&#8203;:abc\)\)

>
> FWIW\, the non\-core Data&#8203;::Dump module does the same thing\.
>

The problem is internal to perl\, so any module that prints it out will
exhibit the same behaviour\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @clintongormley

> I expected both lines to be the same but they differ&#8203;:
> 
> $VAR1 = qr/\(?^u&#8203;:abc\)/;
> 
> $VAR1 = qr/\(?^u&#8203;:\(?^u&#8203;:abc\)\)/;
> 


Rafl had an idea for making this work \(which I have been meaning to turn
into a module\)&#8203;:

Regexp&#8203;::Serialize&#8203;:
=================================================
package Regexp&#8203;::Serialize;

use strict;
use warnings;

our $VERSION = 0\.01;

use Sub&#8203;::Exporter \-setup => \{
&nbsp;    exports => \[ 're\_freeze'\, 're\_thaw'\, 're\_thaw\_eval' \]\,
&nbsp;    groups => \{ default => \[ 're\_freeze'\, 're\_thaw'\, 're\_thaw\_eval' \] \}\,
\};

use XSLoader;
XSLoader&#8203;::load\(\_\_PACKAGE\_\_\);

\#===================================
sub re\_freeze \{ '' \. shift\(\) \}
\#===================================

\#===================================
sub re\_thaw \{
\#===================================
&nbsp;    my $str = shift;
&nbsp;    my $re  = qr/$str/;
&nbsp;    rename\_re\( $re\, $str \);
&nbsp;    return $re;
\}

\#===================================
sub re\_thaw\_eval \{
\#===================================
&nbsp;    use re 'eval';
&nbsp;    re\_thaw\(@&#8203;\_\);
\}

1;


Serialize\.xs
=================================================
\#include "EXTERN\.h"
\#include "perl\.h"
\#include "XSUB\.h"

void
rename\_re\(REGEXP \*rx\, SV \*name\)
\{
&nbsp;  char \*str;
&nbsp;  STRLEN len;

&nbsp;  str = SvPV\(name\, len\);
&nbsp;  SvPVX\(rx\) = str;
&nbsp;  SvCUR\_set\(rx\, len\);
\}

MODULE = Regexp&#8203;::Serialize PACKAGE = Regexp&#8203;::Serialize

void
rename\_re \(SV \*rx\, SV \*name\)
&nbsp;  CODE&#8203;:
&nbsp;    if \(SvTYPE\(SvRV\(rx\)\) \!= SVt\_REGEXP\)
&nbsp;      croak\("not a regular expression"\);

&nbsp;    rename\_re\(\(REGEXP\*\)SvRV\(rx\)\, name\);

**Comment by p5pRT** [comment]
### From @tsee

On 10/23/2012 03&#8203;:59 PM\, Clinton Gormley wrote&#8203;:
>> I expected both lines to be the same but they differ&#8203;:
>>
>> $VAR1 = qr/\(?^u&#8203;:abc\)/;
>>
>> $VAR1 = qr/\(?^u&#8203;:\(?^u&#8203;:abc\)\)/;
>
> Rafl had an idea for making this work \(which I have been meaning to turn
> into a module\)&#8203;:

> package Regexp&#8203;::Serialize;

This should also be doable in Dumper\.pm & Dumper\.xs\. I've known about 
this problem for a while but never had the right combination of time and 
interest to fix it\. The "only" big problem\, I think is making it work 
across the many versions of Perl that D&#8203;::D supports\. This is generally a 
hassle with regular expressions since they traditionally changed a lot\. 
Example&#8203;: Sereal currently fails its tests on 5\.10\.X \(but not before\, not 
after\) because of regexp roundtripping\. And that's despite Yves working 
on it\.\.\.

In a nutshell&#8203;: Patches to Data&#8203;::Dumper would be much appreciated\!

\-\-Steffen

**Comment by p5pRT** [comment]
### From @demerphq

On 23 October 2012 15&#8203;:45\, Brad Gilbert \<b2gills@&#8203;gmail\.com> wrote&#8203;:
> On Tue\, Oct 23\, 2012 at 7&#8203;:28 AM\, Ed Avis \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>> \# New Ticket Created by  "Ed Avis"
>> \# Please include the string&#8203;:  \[perl \#115430\]
>> \# in the subject line of all future correspondence about this issue\.
>> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=115430 >
>>
>>
>> This is a bug report for perl from eda@&#8203;waniasset\.com\,
>> generated with the help of perlbug 1\.39 running under perl 5\.14\.2\.
>>
>>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> \[Please describe your issue here\]
>>
>> Data&#8203;::Dumper&#8203;::Dumper gives a string of code you can eval\(\) to get
>> back the original value\.  But when dumping a regular expression\,
>> some additional ?^u stuff is added to the regexp\.  That would be
>> fine\, except that it gets added every time even if already present\.
>> Hence dumping and evaling a regexp adds another ?^u every time\.
>>
>> The dump\-eval cycle may not give back the exact same string of
>> code you started with \- it's fine for it to translate things to a
>> canonical representation \- but it should be idempotent\.  Evaling
>> a dumped value and then dumping it again should give the same thing\.
>>
>>     use 5\.014;
>>     use warnings;
>>     use strict;
>>     use Data&#8203;::Dumper qw\(Dumper\);
>>     our $VAR1;
>>
>>     my $re = qr/abc/;
>>     my $d = Dumper $re;
>>     my $re2 = eval $d;
>>     my $d2 = Dumper $re2;
>>
>>     say $d;
>>     say $d2;
>>
>> I expected both lines to be the same but they differ&#8203;:
>>
>> $VAR1 = qr/\(?^u&#8203;:abc\)/;
>>
>> $VAR1 = qr/\(?^u&#8203;:\(?^u&#8203;:abc\)\)/;
>
> Add&#8203;:
>
>   say $re;
>   say $re2;
>
> then it prints
>
>   \(?^u&#8203;:abc\)
>   \(?^u&#8203;:\(?^u&#8203;:abc\)\)
>
>>
>> FWIW\, the non\-core Data&#8203;::Dump module does the same thing\.
>>
>
> The problem is internal to perl\, so any module that prints it out will
> exhibit the same behaviour\.

Actually no it is not a core bug\, it is a deficiency in the logic used
by DD and other modules\.  Data&#8203;::Dump&#8203;::Streamer has been doing this
right for years\, and the code to support doing it right was migrated
from DDS to core long ago\. See re&#8203;::regexp\_pattern\(\) in list context\.

Yves



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

On 23 October 2012 15&#8203;:59\, Clinton Gormley \<clint@&#8203;traveljury\.com> wrote&#8203;:
>> I expected both lines to be the same but they differ&#8203;:
>>
>> $VAR1 = qr/\(?^u&#8203;:abc\)/;
>>
>> $VAR1 = qr/\(?^u&#8203;:\(?^u&#8203;:abc\)\)/;
>>
>
>
> Rafl had an idea for making this work \(which I have been meaning to turn
> into a module\)&#8203;:
>
> Regexp&#8203;::Serialize&#8203;:
> =================================================
> package Regexp&#8203;::Serialize;
>
> use strict;
> use warnings;
>
> our $VERSION = 0\.01;
>
> use Sub&#8203;::Exporter \-setup => \{
>     exports => \[ 're\_freeze'\, 're\_thaw'\, 're\_thaw\_eval' \]\,
>     groups => \{ default => \[ 're\_freeze'\, 're\_thaw'\, 're\_thaw\_eval' \] \}\,
> \};
>
> use XSLoader;
> XSLoader&#8203;::load\(\_\_PACKAGE\_\_\);
>
> \#===================================
> sub re\_freeze \{ '' \. shift\(\) \}
> \#===================================
>
> \#===================================
> sub re\_thaw \{
> \#===================================
>     my $str = shift;
>     my $re  = qr/$str/;
>     rename\_re\( $re\, $str \);
>     return $re;
> \}
>
> \#===================================
> sub re\_thaw\_eval \{
> \#===================================
>     use re 'eval';
>     re\_thaw\(@&#8203;\_\);
> \}
>
> 1;
>
>
> Serialize\.xs
> =================================================
> \#include "EXTERN\.h"
> \#include "perl\.h"
> \#include "XSUB\.h"
>
> void
> rename\_re\(REGEXP \*rx\, SV \*name\)
> \{
>   char \*str;
>   STRLEN len;
>
>   str = SvPV\(name\, len\);
>   SvPVX\(rx\) = str;
>   SvCUR\_set\(rx\, len\);
> \}
>
> MODULE = Regexp&#8203;::Serialize PACKAGE = Regexp&#8203;::Serialize
>
> void
> rename\_re \(SV \*rx\, SV \*name\)
>   CODE&#8203;:
>     if \(SvTYPE\(SvRV\(rx\)\) \!= SVt\_REGEXP\)
>       croak\("not a regular expression"\);
>
>     rename\_re\(\(REGEXP\*\)SvRV\(rx\)\, name\);

see re&#8203;::regexp\_pattern\(\)\.

yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

On 23 October 2012 21&#8203;:52\, Steffen Mueller \<smueller@&#8203;cpan\.org> wrote&#8203;:
> On 10/23/2012 03&#8203;:59 PM\, Clinton Gormley wrote&#8203;:
>>>
>>> I expected both lines to be the same but they differ&#8203;:
>>>
>>> $VAR1 = qr/\(?^u&#8203;:abc\)/;
>>>
>>> $VAR1 = qr/\(?^u&#8203;:\(?^u&#8203;:abc\)\)/;
>>
>>
>> Rafl had an idea for making this work \(which I have been meaning to turn
>> into a module\)&#8203;:
>
>
>> package Regexp&#8203;::Serialize;
>
>
> This should also be doable in Dumper\.pm & Dumper\.xs\. I've known about this
> problem for a while but never had the right combination of time and interest
> to fix it\. The "only" big problem\, I think is making it work across the many
> versions of Perl that D&#8203;::D supports\. This is generally a hassle with regular
> expressions since they traditionally changed a lot\. Example&#8203;: Sereal
> currently fails its tests on 5\.10\.X \(but not before\, not after\) because of
> regexp roundtripping\. And that's despite Yves working on it\.\.\.
>
> In a nutshell&#8203;: Patches to Data&#8203;::Dumper would be much appreciated\!

Just copy the logic used by DDS\.

And the only thing holding us up on 5\.10\.x is illness and that I
havent gotten around to building a 5\.10\.x for testing yet\. :\-\)

Yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @tsee

On 10/24/2012 09&#8203;:33 AM\, demerphq wrote&#8203;:
> On 23 October 2012 15&#8203;:45\, Brad Gilbert \<b2gills@&#8203;gmail\.com> wrote&#8203;:
>> The problem is internal to perl\, so any module that prints it out will
>> exhibit the same behaviour\.
>
> Actually no it is not a core bug\, it is a deficiency in the logic used
> by DD and other modules\.  Data&#8203;::Dump&#8203;::Streamer has been doing this
> right for years\, and the code to support doing it right was migrated
> from DDS to core long ago\. See re&#8203;::regexp\_pattern\(\) in list context\.

It is a bug in DD\, not the core\. But rt\.perl\.org is still the 
appropriate place to file bugs\. Data&#8203;::Dumper is maintained as part of 
the core\.

\-\-Steffen

**Comment by p5pRT** [comment]
### From @khwilliamson

On Wed Oct 24 12&#8203;:37&#8203;:42 2012\, smueller@&#8203;cpan\.org wrote&#8203;:
> On 10/24/2012 09&#8203;:33 AM\, demerphq wrote&#8203;:
> > On 23 October 2012 15&#8203;:45\, Brad Gilbert \<b2gills@&#8203;gmail\.com> wrote&#8203;:
> >> The problem is internal to perl\, so any module that prints it out will
> >> exhibit the same behaviour\.
> >
> > Actually no it is not a core bug\, it is a deficiency in the logic used
> > by DD and other modules\.  Data&#8203;::Dump&#8203;::Streamer has been doing this
> > right for years\, and the code to support doing it right was migrated
> > from DDS to core long ago\. See re&#8203;::regexp\_pattern\(\) in list context\.
> 
> It is a bug in DD\, not the core\. But rt\.perl\.org is still the 
> appropriate place to file bugs\. Data&#8203;::Dumper is maintained as part of 
> the core\.
> 
> \-\-Steffen
> 

I looked at Data&#8203;::Dumper\.pm source\, and see this outdated comment&#8203;:
&nbsp;        \# This really sucks\, re&#8203;:regexp\_pattern is in ext/re/re\.xs and
not in 
&nbsp;        \# universal\.c\, and even worse we cant just require that re to be
loaded
&nbsp;        \# we \*have\* to use\(\) it\. 
&nbsp;        \# We should probably move it to universal\.c for 5\.10\.1 and fix this\.
&nbsp;        \# Currently we only use re&#8203;::regexp\_pattern when the re is
blessed into another
&nbsp;        \# package\. This has the disadvantage of meaning that a DD dump
won't round trip
&nbsp;        \# as the pattern will be repeatedly wrapped with the same modifiers\.
&nbsp;        \# This is an aesthetic issue so we will leave it for now\, but we
could use
&nbsp;        \# regexp\_pattern\(\) in list context to get the modifiers separately\.
&nbsp;        \# But since this means loading the full debugging engine in
process we wont
&nbsp;        \# bother unless its necessary for accuracy\.

The reason it is outdated is that re&#8203;::regexp\_pattern has been put in
universal\.c


\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon Nov 12 19&#8203;:59&#8203;:39 2012\, khw wrote&#8203;:
> On Wed Oct 24 12&#8203;:37&#8203;:42 2012\, smueller@&#8203;cpan\.org wrote&#8203;:
> > On 10/24/2012 09&#8203;:33 AM\, demerphq wrote&#8203;:
> > > On 23 October 2012 15&#8203;:45\, Brad Gilbert \<b2gills@&#8203;gmail\.com> wrote&#8203;:
> > >> The problem is internal to perl\, so any module that prints it out will
> > >> exhibit the same behaviour\.
> > >
> > > Actually no it is not a core bug\, it is a deficiency in the logic used
> > > by DD and other modules\.  Data&#8203;::Dump&#8203;::Streamer has been doing this
> > > right for years\, and the code to support doing it right was migrated
> > > from DDS to core long ago\. See re&#8203;::regexp\_pattern\(\) in list context\.
> > 
> > It is a bug in DD\, not the core\. But rt\.perl\.org is still the 
> > appropriate place to file bugs\. Data&#8203;::Dumper is maintained as part of 
> > the core\.
> > 
> > \-\-Steffen
> > 
> 
> I looked at Data&#8203;::Dumper\.pm source\, and see this outdated comment&#8203;:
>         \# This really sucks\, re&#8203;:regexp\_pattern is in ext/re/re\.xs and
> not in 
>         \# universal\.c\, and even worse we cant just require that re to be
> loaded
>         \# we \*have\* to use\(\) it\. 
>         \# We should probably move it to universal\.c for 5\.10\.1 and fix this\.
>         \# Currently we only use re&#8203;::regexp\_pattern when the re is
> blessed into another
>         \# package\. This has the disadvantage of meaning that a DD dump
> won't round trip
>         \# as the pattern will be repeatedly wrapped with the same modifiers\.
>         \# This is an aesthetic issue so we will leave it for now\, but we
> could use
>         \# regexp\_pattern\(\) in list context to get the modifiers separately\.
>         \# But since this means loading the full debugging engine in
> process we wont
>         \# bother unless its necessary for accuracy\.
> 
> The reason it is outdated is that re&#8203;::regexp\_pattern has been put in
> universal\.c
> 

Steffen\, Karl\, Yves&#8203;:  Can we get an update on the status of this ticket?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @tonycoz

On Fri\, Dec 13\, 2013 at 04&#8203;:45&#8203;:46PM \-0800\, James E Keenan via RT wrote&#8203;:
> On Mon Nov 12 19&#8203;:59&#8203;:39 2012\, khw wrote&#8203;:
> > On Wed Oct 24 12&#8203;:37&#8203;:42 2012\, smueller@&#8203;cpan\.org wrote&#8203;:
> > > On 10/24/2012 09&#8203;:33 AM\, demerphq wrote&#8203;:
> > > > On 23 October 2012 15&#8203;:45\, Brad Gilbert \<b2gills@&#8203;gmail\.com> wrote&#8203;:
> > > >> The problem is internal to perl\, so any module that prints it out will
> > > >> exhibit the same behaviour\.
> > > >
> > > > Actually no it is not a core bug\, it is a deficiency in the logic used
> > > > by DD and other modules\.  Data&#8203;::Dump&#8203;::Streamer has been doing this
> > > > right for years\, and the code to support doing it right was migrated
> > > > from DDS to core long ago\. See re&#8203;::regexp\_pattern\(\) in list context\.
> > > 
> > > It is a bug in DD\, not the core\. But rt\.perl\.org is still the 
> > > appropriate place to file bugs\. Data&#8203;::Dumper is maintained as part of 
> > > the core\.
> > > 
> > > \-\-Steffen
> > > 
> > 
> > I looked at Data&#8203;::Dumper\.pm source\, and see this outdated comment&#8203;:
> >         \# This really sucks\, re&#8203;:regexp\_pattern is in ext/re/re\.xs and
> > not in 
> >         \# universal\.c\, and even worse we cant just require that re to be
> > loaded
> >         \# we \*have\* to use\(\) it\. 
> >         \# We should probably move it to universal\.c for 5\.10\.1 and fix this\.
> >         \# Currently we only use re&#8203;::regexp\_pattern when the re is
> > blessed into another
> >         \# package\. This has the disadvantage of meaning that a DD dump
> > won't round trip
> >         \# as the pattern will be repeatedly wrapped with the same modifiers\.
> >         \# This is an aesthetic issue so we will leave it for now\, but we
> > could use
> >         \# regexp\_pattern\(\) in list context to get the modifiers separately\.
> >         \# But since this means loading the full debugging engine in
> > process we wont
> >         \# bother unless its necessary for accuracy\.
> > 
> > The reason it is outdated is that re&#8203;::regexp\_pattern has been put in
> > universal\.c
> > 
> 
> Steffen\, Karl\, Yves&#8203;:  Can we get an update on the status of this ticket?

This looks like a duplicate of \#82948\, which I'll patch \(though
perhaps not a final patch\) this week\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

On Sat Dec 14 06&#8203;:16&#8203;:28 2013\, tonyc wrote&#8203;:
> On Fri\, Dec 13\, 2013 at 04&#8203;:45&#8203;:46PM \-0800\, James E Keenan via RT wrote&#8203;:
> > On Mon Nov 12 19&#8203;:59&#8203;:39 2012\, khw wrote&#8203;:
> > > On Wed Oct 24 12&#8203;:37&#8203;:42 2012\, smueller@&#8203;cpan\.org wrote&#8203;:
> > > > On 10/24/2012 09&#8203;:33 AM\, demerphq wrote&#8203;:
> > > > > On 23 October 2012 15&#8203;:45\, Brad Gilbert \<b2gills@&#8203;gmail\.com>
> > > > > wrote&#8203;:
> > > > >> The problem is internal to perl\, so any module that prints it
> > > > >> out will
> > > > >> exhibit the same behaviour\.
> > > > >
> > > > > Actually no it is not a core bug\, it is a deficiency in the
> > > > > logic used
> > > > > by DD and other modules\.  Data&#8203;::Dump&#8203;::Streamer has been doing
> > > > > this
> > > > > right for years\, and the code to support doing it right was
> > > > > migrated
> > > > > from DDS to core long ago\. See re&#8203;::regexp\_pattern\(\) in list
> > > > > context\.
> > > >
> > > > It is a bug in DD\, not the core\. But rt\.perl\.org is still the
> > > > appropriate place to file bugs\. Data&#8203;::Dumper is maintained as
> > > > part of
> > > > the core\.
> > > >
> > > > \-\-Steffen
> > > >
> > >
> > > I looked at Data&#8203;::Dumper\.pm source\, and see this outdated comment&#8203;:
> > >         \# This really sucks\, re&#8203;:regexp\_pattern is in ext/re/re\.xs
> > > and
> > >  not in
> > >         \# universal\.c\, and even worse we cant just require that re
> > > to be
> > > loaded
> > >          \# we \*have\* to use\(\) it\.
> > >         \# We should probably move it to universal\.c for 5\.10\.1 and
> > > fix this\.
> > >         \# Currently we only use re&#8203;::regexp\_pattern when the re is
> > > blessed into another
> > >         \# package\. This has the disadvantage of meaning that a DD
> > > dump
> > > won't round trip
> > >         \# as the pattern will be repeatedly wrapped with the same
> > > modifiers\.
> > >         \# This is an aesthetic issue so we will leave it for now\,
> > > but we
> > > could use
> > >         \# regexp\_pattern\(\) in list context to get the modifiers
> > > separately\.
> > >         \# But since this means loading the full debugging engine in
> > > process we wont
> > >         \# bother unless its necessary for accuracy\.
> > >
> > > The reason it is outdated is that re&#8203;::regexp\_pattern has been put
> > > in
> > > universal\.c
> > >
> >
> > Steffen\, Karl\, Yves&#8203;:  Can we get an update on the status of this
> > ticket?
> 
> This looks like a duplicate of \#82948\, which I'll patch \(though
> perhaps not a final patch\) this week\.

The fix for 82948 fixes this\, so closing\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @epa

Thanks for fixing this\.  Both lines now print

$VAR1 = qr/abc/u;
